cli/src/commands/log.rs

```diff
@@ -180,7 +182,15 @@ pub(crate) fn cmd_log(
                 let node_symbol = if Some(&key.0) == wc_commit_id {
                     "@"
                 } else {
-                    &default_node_symbol
+                    let commit_revset = RevsetExpression::commit(key.0.clone());
+                    let revset = workspace_command
+                        .evaluate_revset(commit_revset.intersection(&immutable_revset))?;
+                    let is_immutable = revset.iter().commits(store).next().is_some();
+                    if is_immutable {
+                        &immutable_node_symbol
+                    } else {
+                        &default_node_symbol
+                    }
```

Comment:
This seems a bit expensive. How much does it slow down `jj log -r ..` in the jj repo, for example?

We have talked about adding a `.contains()` method to revsets. The most recent discussion was here:
https://discord.com/channels/968932220549103686/969291218347524238/1213888562790006824

Let me know if you prefer not to log on to Discord and I can paste that text here.

Reply:
Yes, it's definitely slower:

```
λ time jj log -r .. > /dev/null

________________________________________________________
Executed in  546.09 millis    fish           external
   usr time  524.18 millis    0.09 millis  524.10 millis
   sys time   27.16 millis    1.40 millis   25.76 millis
```

versus `main`:

```
λ time jj log -r .. > /dev/null

________________________________________________________
Executed in  134.47 millis    fish           external
   usr time  114.93 millis   78.00 micros  114.85 millis
   sys time   24.40 millis  660.00 micros   23.74 millis
```

> Let me know if you prefer not to log on to Discord and I can paste that text here.

Could you please paste the text here? I would greatly appreciate it.

Reply:
Here you go:

> fyi, the main complication is that in order to make it fast, we'll want to be able to ask revset if it contains a given commit. we'd ask the ancestors(immutable_heads()) revset if it contains the commit we're about to render. for the default revset engine (the one bundled by default), we will do that by walking the revset incrementally while caching the set of commits visited. for the revset engine at google, we'll be able to answer it more efficiently (thankfully, because it's not feasible to walk all ancestors). i think we know what we need to do. i just haven't gotten around to it

> it'll be someting like let immutable_revset = immutable_revset.caching(); for commit_id in revset_to_draw.iter() { let is_immutable = immutable_revset.contains(&commit_id); <render it>; } 

We have the current `Revset` trait and we would add a `caching()` `with_contains()`, something like this:

```rust
trait Revset {
  fn iter(&self) -> Box<dyn Iterator<Item = CommitId> + '_>;
  // etc

  //
  fn caching(&self) -> Box<dyn CachingRevsetIterator + '_>;
}


trait CachingRevsetIterator : Iterator<Item = CommitId> {

  fn contains(&self, id: &CommitId) -> bool;
}
```

Reply:
Thank you! I'll look into it later.

Reply:
> fn caching(&self) -> Box<dyn CachingRevsetIterator + '_>;

fwiw, we might have to reintroduce the `'index` lifetime to enable `immutable.contains(commit_id)` template keyword. For `jj log` use case, it wouldn't be needed. `+ '_` should be good enough.

Reply:
Ok. I'm not sure I understand what implementation of `CachingRevsetIterator` should be. Is the following solution (pseudocode) would work?
```rust
struct CachingRevsetIteratorImpl {
    index: CompositeIndex,
    entries: Box<dyn Iterator<Item = IndexEntry>,    
    visited_commit_ids: HashSet<CommitId>,
    visited_pos: Option<IndexPosition>,
}

impl CachingRevsetIteratorImpl {
    fn next(&mut self) -> Option<CommitId> {
        // Consume revset incrementally while caching each commit.
        if let Some(entry) in self.entries.next() {
            visited_commit_ids.insert(entry.commit_id());
            visited_pos = entry.position();
            return Some(entry.commit_id())
        }
        None
    }

    fn contains(&mut self, commit_id: &CommitId) -> bool {
        if visited_commit_ids.contains(commit_id) {
            return true;
        }

        while let Some(visited_commit_id) = self.next() {
            if visited_commit_id == commit_id  {
                return true;
            }

            match (self.index.commit_id_to_pos(commit_id), self.visited_pos) {
                (Some(commit_id_pos), Some(visited_pos)) => {
                    // There is no point going down, since the commit would already 
                    // have been visited at this stage.
                    if commit_id_pos > visited_pos {
                        return false;
                    }
                }
                _ => continue;
            }

        }

        visited_commit_ids.contains(id)
    }
}
```

Reply:
Yes, exactly! Except that `next()` would have to first return any items previously peeked at by `contains()`.

Reply:
Perhaps, the source iterator and the cached positions would have to be shared across iterators (via `Rc<RefCell<..>>` or something)? (I assume the cached wrapper will provide `.iter()` to start over.)

```rust
CachedRevset {
  fn iter(&self) -> Box<dyn Iterator<Item = Commit>>;
  fn contains(&self, &CommitId) -> bool;
}
```
The `IndexPosition` returned from the source iterator is sorted, so `contains()` can do binary search over `Vec<IndexPosition>`.

Reply:
> it would be a different iterator type rather than a different revset type

It doesn't have to be, but the cache object would have `{ visited: HashMap<IndexPosition>, nexts: VecDeque<IndexPosition> }` or `{ visited: Vec<IndexPosition>, next_pos: usize }` anyway, so adding separate `.iter()` wouldn't complicate things.